Mathlib/Analysis/Normed/Affine/AsymptoticCone.lean

```diff
@@ -0,0 +1,428 @@
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem asymptoticCone_union {s t : Set P} :
+    asymptoticCone k (s ∪ t) = asymptoticCone k s ∪ asymptoticCone k t := by
+  ext
+  simp only [Set.mem_union, mem_asymptoticCone_iff, Filter.frequently_or_distrib]
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem Set.Finite.asymptoticCone_biUnion {ι : Type*} {s : Set ι} (hs : s.Finite) (f : ι → Set P) :
+    asymptoticCone k (⋃ i ∈ s, f i) = ⋃ i ∈ s, asymptoticCone k (f i) := by
+  induction s, hs using Set.Finite.induction_on <;>
+    simp [asymptoticCone_union, *]
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem Set.Finite.asymptoticCone_sUnion {S : Set (Set P)} (hS : S.Finite) :
+    asymptoticCone k (⋃₀ S) = ⋃ s ∈ S, asymptoticCone k s := by
+  rw [sUnion_eq_biUnion, hS.asymptoticCone_biUnion]
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem Finset.asymptoticCone_biUnion {ι : Type*} (s : Finset ι) (f : ι → Set P) :
+    asymptoticCone k (⋃ i ∈ s, f i) = ⋃ i ∈ s, asymptoticCone k (f i) :=
+  s.finite_toSet.asymptoticCone_biUnion f
+
+omit [OrderTopology k] [IsStrictOrderedRing k] in
+theorem asymptoticCone_iUnion_of_finite {ι : Type*} [Finite ι] (f : ι → Set P) :
+    asymptoticCone k (⋃ i, f i) = ⋃ i, asymptoticCone k (f i) := by
+  rw [← Set.sUnion_range, (Set.finite_range _).asymptoticCone_sUnion, Set.biUnion_range]
+
+theorem zero_mem_asymptoticCone {s : Set P} (hs : s.Nonempty) : 0 ∈ asymptoticCone k s := by
+  simpa [mem_asymptoticCone_iff]
+
+theorem asymptoticCone_nonempty_iff {s : Set P} : (asymptoticCone k s).Nonempty ↔ s.Nonempty := by
+  refine ⟨Function.mtr ?_, fun h => ⟨0, zero_mem_asymptoticCone h⟩⟩
+  simp +contextual [Set.not_nonempty_iff_eq_empty]
+
+theorem smul_mem_asymptoticCone_iff {s : Set P} {c : k} {v : V} (hc : 0 < c) :
+    c • v ∈ asymptoticCone k s ↔ v ∈ asymptoticCone k s := by
+  simp_rw [mem_asymptoticCone_iff, asymptoticNhds_smul v hc]
+
+theorem smul_mem_asymptoticCone {s : Set P} {c : k} {v : V} (hc : 0 ≤ c)
+    (h : v ∈ asymptoticCone k s) : c • v ∈ asymptoticCone k s := by
+  rcases hc.eq_or_lt with rfl | hc
+  · rw [zero_smul]; exact zero_mem_asymptoticCone (asymptoticCone_nonempty_iff.mp ⟨v, h⟩)
+  · rwa [smul_mem_asymptoticCone_iff hc]
+
+theorem asymptoticCone_affineSubspace {s : AffineSubspace k P} (hs : (s : Set P).Nonempty) :
+    asymptoticCone k (s : Set P) = closure s.direction := by
+  have ⟨p, hp⟩ := hs
+  ext v
+  rw [mem_closure_iff_frequently, ← map_snd_prod (atTop (α := k)) (𝓝 v), frequently_map,
+    mem_asymptoticCone_iff, asymptoticNhds_eq_smul_vadd v p, vadd_pure, frequently_map, ← map₂_smul,
+    ← map_prod_eq_map₂, frequently_map]
+  apply frequently_congr
+  filter_upwards [tendsto_fst.eventually (eventually_ne_atTop 0)] with ⟨c, u⟩ hc
+  simp_rw [SetLike.mem_coe, s.vadd_mem_iff_mem_direction _ hp, Submodule.smul_mem_iff _ hc]
+
+theorem asymptoticCone_submodule {s : Submodule k V} :
+    asymptoticCone k (s : Set V) = closure s := by
```

Comment:
```suggestion
@[simp] theorem asymptoticCone_submodule (s : Submodule k V) :
    asymptoticCone k (s : Set V) = closure s := by
```
and can you generalise this to `ConvexConeClass`?

Reply:
I think this shouldn't be `simp`, because it causes non-confluence with `Submodule.inf_coe`.

There's currently no `ConvexConeClass`. Instead, I added a more general variant that can be used with `ConvexCone`.